### ISSUE TITLE
'Pending communication' not defined in definition of MPI_Comm_disconnect

### ISSUE THREAD

**Issue by RolfRabenseifner** [mpi-forum/mpi-issues#710]
# Problem

The text uses the term "pending communication". Nowhere is it defined. Here, "decoupled MPI activities" may be the correct wording, see new MPI-4.1 Section 2.9 Progress.

§11.10.4
> `MPI_COMM_DISCONNECT` has the same action as `MPI_COMM_FREE`, except that it waits for *pending communication* to finish internally and enables the guarantee about the behavior of disconnected processes.

would be

> `MPI_COMM_DISCONNECT` has the same action as `MPI_COMM_FREE`, except that it waits for *decoupled MPI activities* to finish internally and enables the guarantee about the behavior of disconnected processes.

Another problem is, that the the phrase "has the same action ... except ..." is wrong, because the
- requirements for calling the routine are quite different,
  which implies also different outcome (=action) if the requirements are not fulfilled, and
- the outcome for inactive request handles on the  given communicator is also different,
  see also https://github.com/mpi-forum/mpi-issues/issues/536.

Furthermore, the text in MPI-4.0

> MPI_COMM_DISCONNECT may be called only if all communication is complete and matched, so that buffered data can be delivered to its destination. This requirement is the same as for MPI_FINALIZE.

seems not to be clear enough, as the discussion at the MPI forum meeting May 3, 2023 has shown.

Especially in the definition of MPI_Finalize

>Before an MPI process invokes MPI_FINALIZE, the process must perform all MPI calls needed to complete its involvement in MPI communications associated with the World Model. It must locally complete all MPI operations that it initiated and must execute matching calls needed to complete MPI communications initiated by other processes. For example, if the process executed a nonblocking send, it must eventually call MPI_WAIT, MPI_TEST, MPI_REQUEST_FREE, or any derived function; if the process is the target of a send, then it must post the matching receive; if it is part of a group executing a collective operation, then it must have completed its participation in the operation.

it is clear that the calls to MPI_WAIT, MPI_TEST, or MPI_REQUEST_FREE (to free request handles of blocking operations or to complete or free persistent requests) nmust be called before MPI_Finalize.

Open questions have been
- whether the *eventually* only means that they must be called between the starting of the request and MPI_Finalize or whether it has another or additional meaning?
- and whether in the case of MPI_COMM_DISCONNECT, what is allowed after MPI_COMM_DISCONNECT returned in the case of inactive persistent request handles?
- whether in the future, this list may be extended by adding MPI_REQUEST_GET_STATUS (and its derived functions),
  and which special rules then apply for the still existing request handle?

Consensus was that MPI_Finalize nor the sections about pt-to-pt and collective persistent operations require that an inactive persistent request handle is freed  with MPI_REQUEST_FREE before calling MPI_Finalize.

See also discussions in issues https://github.com/mpi-forum/mpi-issues/issues/543 and https://github.com/mpi-forum/mpi-issues/issues/676

# Proposal

The proposal in the pull request tries onyl to resolve the inconsistencies, but as this is an errata, not any future extensions for MPI_REQUEST_GET_STATUS.
In addition, it makes the same clarification for MPI_Finalize. 

# Changes to the Text

See above. This should be part of the "cleanup" for MPI 4.1.

Alternatively, the term "pending communication" could be properly in the terms section.
(Risky, because its meaning may be different in different contexts)

# Impact on Implementations

None (unless there it renders some implementations non-compliant due to confusion stemming from the usage of the term "pending") 
Exception: If an MPI implementation requires that related persistent requests are freed (instead of only beeing inactive) before calling MPI_COMM_DISCONNECT (i.e. requiring more than for MPI_FINALIZE), then this bug should be fixed.

# Impact on Users

Clear use of terms helps users better understand the text without having to reason about what a "pending operation" is.

# References and Pull Requests

[PR 823](https://github.com/mpi-forum/mpi-standard/pull/823)

**Comment by RolfRabenseifner** [comment]
@hppritcha @raffenet  @Wee-Free-Scot @schulzm @naughtont3 / @GuillaumeMercier / @devreal @bkmgit 
I assigned you all as member of the process-start-chapter-committee, or as groups-chapter-chair, or as active yesterday in the discussion-chat.

I tried to summarize the problems here on the abstract of this issue.
I tried to produce a minimal consistent solution that I would propose. 
I explicitly refused to add anything about MPI_REQUEST_GET_STATUS, 
because the corner case positiv test with MPI_REQUEST_GET_STATUS plus no call to MPI_WAIT/TEST/COMM_FREE
is not yet handled in the rules for MPI_FINALIZE and therefore not part of this errata ticket.

In my opinion, it is entirely new territory to promote dangling "request handles" that are never released.

I hope you like the content of the proposal. I feel the rules must be clear for the users - that should be an important goal.

Issue: https://github.com/mpi-forum/mpi-issues/issues/710
PR: https://github.com/mpi-forum/mpi-standard/pull/823

**Comment by RolfRabenseifner** [comment]
The changes are marked with PR823 and also in yellow color on pages 475, 525, 1025, 1054 in
[mpi41-report_Issue710_PR823.pdf](https://github.com/mpi-forum/mpi-standard/files/11399764/mpi41-report_Issue710_PR823.pdf)

**Comment by bosilca** [comment]
If we only require persistent communications to be inactive, we are forcing a high-quality (aka memory-leak free) MPI implementation to track all persistent requests to be able to release them (on behalf of the user) during finalization.

Moreover, having different requirements for nonblocking (which are required to be freed) and persistent requests (which are required to be inactive **or** freed) is not helping with the consistency. From an implementor perspective it makes more sense to require that all persistent requests are inactive **and** freed, or simply all requests to be **freed** (and then the part about `decoupled MPI activities` will take care of the rest).

**Comment by RolfRabenseifner** [comment]
For MPI finalize, no MPI library required persistent handles to be freed. We also do not reqire datatype handles or ... to be freed before finalize. Would it be helpful to write an AtoU?

**Comment by bosilca** [comment]
Open MPI does. Our objects refcount each other (datatype to requests, requests to communicators, ...) such that at the end we know if there are dangling objects. In debug mode we even print a list of dangling objects.

**Comment by raffenet** [comment]
MPICH will also print a list of objects still allocated objects during MPI_FINALIZE in a debug build. Though I am not sure the user is _required_ to do anything with that information.

**Comment by Wee-Free-Scot** [comment]
A possible approach to dealing with dangling/stale handles/objects is to permit MPI_Request_free (and MPI_Type_free, ...) to be called *after* MPI_Finalize. I'm not saying it's a good approach, but it points out that "valgrind cleanliness" is not just the responsibility of the MPI library, but is a cooperation between MPI and the user.

Open MPI keeps a list internally of active requests, so it can tell the difference between "there are no active requests right now" and "there is at least one active request right now". It can partition the list of active requests by the session they are derived from, so it can tell the difference between "there are no active requests for session X (which might be the World model)" and "there is at least one active request for session X (which might be the World model)".

During MPI_Finalize/MPI_Session_finalize, even one active request *in the World model/particular session* means an erroneous program (go mad, do whatever you want).

During MPI_Comm_disconnect for a communicator derived from the World/Session model, even one active request *in the World model/particular session* means an erroneous program (go mad, do whatever you want).

The problem highlighted is exclusively about the existence of inactive requests at the point of destroying a communicator (i.e. at the point of calling finalize or disconnect).

Axiom: The internal request structure must keep a pointer to the communicator (not strictly true, but lets pretend that is an axiom and argue about it separately).

Question: why is that pointer contributing to the ref-count on the communicator when the request is inactive?

If MPI_Start (and friends) incremented the ref-count on the communicator and MPI_Wait (and friends) decremented the ref-count on the communicator, then the existence of inactive requests would not contribute to the ref-count and would not prevent MPI from destroying the communicator when told to do so, e.g. by MPI_Comm_disconnect.

The obvious objection is that destroying a request possibly/sometimes involves destroying the objects to which it holds a pointer (i.e. when that pointer is the last ref-counted reference to that other object).

Question: why does destroying a request cascade to potentially destroying the communicator for which it has a pointer?

The answer is MPI_Comm_free -- when MPI_Comm_free only marks the communicator for later destruction (because the ref-count is nonzero), it expects that some later action will actually destroy the communicator (when the ref-count becomes zero).

If inactive requests have a reference to the communicator (as well as a pointer to it), then the existence of even one inactive request will prevent MPI_Comm_free from destroying the communicator. In this case, destroying inactive requests to remove the reference is necessary before the communicator can be destroyed -- which means an inactive request is a _pending communication_ in the sense of the MPI Standard wording for MPI_Comm_disconnect,

If inactive requests have a pointer to the communicator, but not a reference (i.e. it is known by design that the communicator ref-count (if it still exists) is not counting this pointer at this time), then their existence does not prevent MPI_Comm__free from destroying the communicator; only active requests have a reference and can prevent destruction during MPI_Comm_free. In this case, destroying inactive requests is not necessary before the communicator can be destroyed (because they do not contribute to the ref-count) -- which means an inactive request is NOT a _pending communication_ in the sense of the MPI Standard wording for MPI_Comm_disconnect,

Consequences, part 1:

If an inactive request does NOT have a reference to the communicator, then the communicator could be destroyed by MPI while it is inactive, and attempting to start it again after MPI_Comm_free is then going to be a potential problem. MPI_Start might attempt to dereference a dangling communicator pointer and bad things (UB) will likely happen. OTOH, MPI_Comm_free/MPI_Comm_disconnect can safely ignore the existence of inactive requests (MPI does not even need to know whether any inactive requests exist) and destroy the communicator if/when the ref-count (which counts *active* references for _pending communication_ only) is zero.

Thus, if inactive requests are NOT _pending communications_ then
* we permit inactive requests to exist during MPI_Comm_disconnect and MPI_Finalize/MPI_Session_finalize (and those procedures will actually destroy the communicator before they return, as they are supposed to)
* we lose the ability to start inactive requests after MPI_Comm_free, which is currently allowed by omission in the MPI Standard, works in some MPI libraries, and is used by some customers.

Consequences, part 2:

If an inactive request DOES have a reference to the communicator, then the communicator cannot be destroyed while an inactive request exists; attempting to start it again after MPI_Comm_free is then going to be fine (probably). MPI_Start would attempt to dereference the communicator pointer and good things will likely happen. OTOH, MPI_Comm_free/MPI_Comm_disconnect must not ignore the existence of inactive requests and must not destroy the communicator while they still exist, so they must contribute to the ref-count that triggers/prevents the communicator destruction.

Thus, if inactive requests ARE _pending communications_ then
* we permit inactive requests to be used normally after MPI_Comm_free (but not after MPI_Comm_disconnect, which should hang until these requests are destroyed)
* we lose the ability to destroy the communicator whilst inactive requests exist, which is currently allowed by omission in the MPI Standard, works in some MPI libraries, and is used by some customers.

Conclusion

We have a choice: which group of customers do we disappoint? Those who want to use (start, wait, etc) inactive requests after freeing the communicator handle or those who want to destroy the communicator itself without freeing inactive requests beforehand?

Which is all a really long way of getting to the suggestion of permitting inactive requests to be freed *after* the communicator is destroyed. If "the request is inactive" means "has a pointer, but not a reference, to a communicator" then MPI_Request_free must not follow that pointer, it just destroys the request itself, treating the communicator pointer as if it were dangling (even if it is not). That can happen before the communicator is destroyed (the pointer would work, if used, but no-one cares because there must be at least one other reference to the communicator so MPI_Request_free has no work to do) or afterwards (the pointer is dangling but no-one cares because the communicator is already gone so MPI_Request_free has no work to do).

To achieve valgrind cleanliness, the user must free all their requests and destroy all their communicators, of course, *but not necessarily in that order*.

We now have an option that doesn't disappoint either group of customers, but involves re-work in some MPI implementations.

**Comment by raffenet** [comment]
So my view of the options are:
1. _Require_ the user to free inactive requests before disconnect/finalize
2. Allow the user to free inactive requests after disconnect/finalize (or not and get a warning from debug builds)
3. Allow _MPI_ to free inactive requests _during_ disconnect/finalize, thus allowing the user to free them, or not, before disconnect/finalize without warning. This requires some additional state tracking by the MPI library, but only for persistent requests allocations off the critical path. This option seemed palatable when discussing within the MPICH team, especially in the context of the sessions model.

**Comment by Wee-Free-Scot** [comment]
> 1. _Require_ the user to free inactive requests before disconnect/finalize

Well ... "before disconnect/finalize *can return*".

With thread multiple support, the user can free inactive requests on a different thread to release the thread blocked on disconnect/finalize. Without thread multiple support, then yes exactly what Ken wrote because my wording and Ken's are equivalent in that situation.

**Comment by RolfRabenseifner** [comment]
> 1. _Require_ the user to free inactive requests before disconnect/finalize

Finalize and persistent request exist since MPI-1 and  freeing before finalize was never required.
Invalidating all existing applications that use persistent requests and not freeing them before finalize would be a no-go for me.

> 2. Allow the user to free inactive requests after disconnect/finalize (or not and get a warning from debug builds)

After the discussion, I would really prefer Option 3:

> 3. Allow _MPI_ to free inactive requests _during_ disconnect/finalize, thus allowing the user to free them, or not, before disconnect/finalize without warning. This requires some additional state tracking by the MPI library, but only for persistent requests allocations off the critical path. This option seemed palatable when discussing within the MPICH team, especially in the context of the sessions model.

This is for me the best option. It is very clear and symmetric:
- We did never require the freeing of inactive persistent requests but we allow it and this feature is kept.
- It is symmetric for finalize and disconnect: in both cases, i.e., after finalize and disconnect, the use of an inactive (=not freed) persistent request handle is erroneous.
- And there is no difference after disconnect in the use of inactive persistent requests: both, MPI_Request_free and MPI_Start are erroneous, i.e., forbidden. 

If there is some consensus on this 3rd option, I would change the pull request accordingly.

**Comment by RolfRabenseifner** [comment]
In addition to my comment above on handling of _**inactive persistent request**_, here, I want to comment on _**nonblocking requests**_:

The following example is correct since MPI-2.0, although it might be seen as a corner case.

``` C
int snd_buf, rcv_buf;
MPI_Init();
MPI_Buffer_attach(...);
MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
MPI_Comm_dup(MPI_COMM_WORLD, &my_comm);
// same ranks on my_comm and MPI_COMM_WORLD
if (my_rank==0) {
  snd_buf = 5;
  MPI_Bsend(snd_buf, 1, MPI_INT, dest=1, tag=0, comm=my_comm);
} else if (my_rank==1) {
  MPI_Request rq;
  MPI_Irecv(rcv_buf, 1, MPI_INT, source=0, tag=0, comm=my_comm, request=&rq);
  MPI_Request_free(&rq); // rq is set to MPI_REQUEST_NULL, but operation need not to be complete:
                         // the rcv_buf must not be read or reused bofore the operation is complete 
}
MPI_Comm_disconnect(&my_comm); // collective over whole my_comm
          // all pending communication is complete
          // i.e.,  - the Irecv has filled in the rcv_buf
          //        - the Bsend must have send out its buffered data
if (my_rank==0) {
  // the application can now reuse the attached buffer for other purposes
} else if (my_rank==1) {
  // the rcv_buf is now filled with the value 5 and can be reused for other purposes
}
MPI_Finalize();
```
Agreed? 

An even more extreme corner case example based on the advice to implementors for MPI_Finalze: 

> "The MPI implementation must ensure that a process has completed any involvement in MPI communication before MPI_FINALIZE returns."

``` C
int snd_buf, rcv_buf;
MPI_Init();
MPI_Buffer_attach(...);
MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
if (my_rank==1) {
  snd_buf = 5;
  MPI_Bsend(snd_buf, 1, MPI_INT, dest=0, tag=0, comm=MPI_COMM_WORLD);
} else if (my_rank==0) {
  MPI_Request rq;
  MPI_Irecv(rcv_buf, 1, MPI_INT, source=1, tag=0, comm=MPI_COMM_WORLD, request=&rq);
  MPI_Request_free(&rq); // rq is set to MPI_REQUEST_NULL, but operation need not to be complete:
                         // the rcv_buf must not be read or reused bofore the operation is complete 
}
MPI_Finalize(); 
          // all pending communication is complete
          // i.e.,  - the Irecv has filled in the rcv_buf
          //        - the Bsend must have send out its buffered data
if (my_rank==1) { // if this process still exists after MPI_Finalize
  // the application can now reuse the attached buffer for other purposes
} else if (my_rank==0) { // rank 0 in MPI_COMM_WORLD must still exist after MPI_Finalize
  // the rcv_buf is now filled with the value 5 and can be reused for other purposes
}
```
Agreed?

**Comment by Wee-Free-Scot** [comment]
In the example codes from @RolfRabenseifner that use buffered mode send operations, there is no attempt to detach the buffer before the call to disconnect/finalize -- do we wish "communication is complete" to mean "there are no internal references to the communicator used in the buffered mode send operations that were using space in the attached buffer"? That is, disconnect/finalize must do all decoupled activities, which involves draining the attached buffer, which is all/most of the work of detaching the buffer, so the user can refuse to detach the buffer but still expect a valgrind clean state at the end. When can the user free the heap-allocated buffer space they attached to MPI?

Aside: can I do a buffered mode send from within an delete attribute callback that is called during disconnect/finalize -- is the buffer still guaranteed to be attached at the point in execution of disconnect/finalize when those callbacks are called by MPI?

After the new proposal that permits a buffered mode buffer to be attached directly to the communicator, do we need to state that disconnect/finalize both do the work of detaching the buffer, if one is still attached? Of course, finalizing a session will also do the work of detaching the buffer attached to the session, if any, as well as doing the work of detaching all the buffers attached to all the communicators derived from that session, thereby releasing all the internal references to those communicators from the decoupled activities related to all the buffered mode send operations that used any of the communicators within that session. Do we need to state that too? Is it adequately covered by "decoupled activities"?

**Comment by Wee-Free-Scot** [comment]
Option 3 in the summary from @raffenet gives me the heebie-jeebies. It's either the status quo or a dangerous security flaw.

``` C
MPI_Recv_init(..., comm, &req);
MPI_Comm_disconnect(&comm);
assert(req == MPI_REQUEST_NULL); // this assert should fail, even if the request has been destroyed
```

In the above code, lets assume we have option (3) with an implementation that chooses to avail itself of this new(?) freedom to "free inactive requests during disconnect/finalize".

1. MPI destroys the request object/struct/whatever that is internal to MPI. The allocation and freeing of internal request objects/structs/whatevers is (and has always been) entirely the purview of MPI. So, that bit seems fine to me -- indeed I would argue that is NOT a new freedom -- it is known during disconnect/finalize that usage of requests that reference any communicator that is being destroyed will be A Bad Idea (tm) and destroying the request is unlikely to make things worse.

2. After disconnect/finalize, EITHER the user has a stale request handle (not MPI_REQUEST_NULL but also not valid) OR somehow MPI set the user's variable (if it still exists) to MPI_REQUEST_NULL. Neither of these is a great situation. The best thing the user can do is let their request handle variable go out of scope without touching it.
    * How does MPI know that scribbling over a user-space memory location (that *was* valid some time ago, but which might not be now) is not A Bad Thing (tm)? Short answer: it doesn't, so I'm strongly against this idea. The assert in the code above MUST fail; the user's variable must not be scribbled over by MPI because it might have gone out of scope by the time MPI gets around to scribbling on it.
    * How does MPI protect itself against all the other possible user actions? Simples -- we can state that usage of request handles after disconnect/finalize is erroneous and punt all problems into a giant SEP field.

I think MPI libraries are permitted this level of freedom already -- this is the status quo (except for an explicit "subsequent usage of request handles is erroneous" statement). An MPI library that chooses NOT to destroy inactive requests during disconnect/finalize, and thereby leak MPI-allocated memory, is just not high quality.

Aside: this makes `MPI_Comm_free` very different, semantically, from disconnect/finalize. Is it still true that `MPI_Comm_free` is "kind of like `MPI_Comm_idisconnect` with a request that gets completed by MPI sometime after the notional communicator ref-count hits zero and before finalize returns"? No. Calling `MPI_Comm_disconnect` gives MPI permission to begin destroying inactive requests, so calling `MPI_Comm_idisconnect` would too. However, the users expect to be able to use request handles that were inactive at the point of calling `MPI_Comm_free` after that call returns, so we now have a race.

**Comment by RolfRabenseifner** [comment]
@Wee-Free-Scot 
> That is, disconnect/finalize must do all decoupled activities, which involves draining the attached buffer, which is all/most of the work of detaching the buffer, so the user can refuse to detach the buffer but still expect a valgrind clean state at the end. When can the user free the heap-allocated buffer space they attached to MPI?

With the new issue https://github.com/mpi-forum/mpi-issues/issues/586, it is the work of MPI_BUFFER_FLUSH (instead of detach+attach).

My comment
>   // the application can now reuse the attached buffer for other purposes

meant that the attached global buffer can be reused for Bsends on other communicators. My comment did not want to say that the buffer is detached.

This means, here I do not want to discuss some automatic detach of the global buffer. 
Furthermore, I would not expect that such automatic detach was ever intended, because also MPI_Finalize does not free any object.

I would really add somthing like

> After calling MPI_COMM_DISCONNECT, any subsequent usage of an inactive persistent request handle is erroneous if this request was created with an operation on comm.

(It is always up to the programmer to fulfill such rules.) 

Okay?

**Comment by bosilca** [comment]
I don't see the point of constantly bringing up `MPI_Finalize`. After the call to `MPI_Finalize` no MPI operations related to communications are allowed, which means nobody cares about the state of the previously allocated handles. This is at the opposite of `MPI_Comm_free` and `MPI_Comm_disconnect` where the MPI library is still up and running, but the state of previously valid handles is not clearly defined.

If we expect MPI libraries to track the persistent requests handles or check that these handles are valid before allowing them to be started, we are imposing an unnecessary overhead on MPI implementors and a waste on users. I think the most straightforward to explain approach is to require all user-facing MPI objects related to the communicator to be freed/released prior to calls to `MPI_Comm_disconnect` and `MPI_Comm_free`. Such an approach keep the execution  path clean and fast (because no side effects can exists), make sense from users and implementors perspective and can be easily implemented with a refcount (and most of the MPI implementations already do this).

**Comment by RolfRabenseifner** [comment]
I expect that the new version is now consistent and has -- based on the discussed examples above  -- not a significant performance drawback. There may be one additional entry in the internal data structure of the MPI library during the init of a persistent request and some additional specific effort during MPI_COMM_DISCONNECT and MPI_SESSION_FINALIZE.

@bosilca There is no aditional overhead with MPI_FINALIZE, but the wording now explicitely allows that a user can free all memory that he/she malloced for the send and receive buffer for the persitent requests. Without this statement, the persistent request is not freed and the MPI standard says that these resources are blocked. This is somehow stupid and fixed through this simple addition cost-free sentence.

All changes of [PR 823](https://github.com/mpi-forum/mpi-standard/pull/823) are marked with PR823 and also in yellow color on pages 100, 475, 525, 1025, 1054, 1057 in
[mpi41-report_Issue710_PR823.pdf](https://github.com/mpi-forum/mpi-standard/files/11454896/mpi41-report_Issue710_PR823.pdf)

**Comment by RolfRabenseifner** [comment]
@bosilca 
>  I think the most straightforward to explain approach is to require all user-facing MPI objects related to the communicator to be freed/released prior to calls to MPI_Comm_disconnect and MPI_Comm_free. 

This would be a non-backward compatible change.
The proposed PR is backward compatible, because it only defines something about the use of associated inactive request handles after MPI_DISCONNECT, MPI_SESSION_FINALIZE and MPI_FINALIZE. It is backward compatible, because until MPI-4.0 the behavior was undefined.

**Comment by bosilca** [comment]
We're turning in circles here without any constructive outcome. Let me give you two examples why this is bad practice:

1. pinned memory. One, extremely good, reason we have persistent requests it to remove the memory pinning from the critical path. We can do the pinning in the persistent requests initialization, and leave it pinned until the persistent request is freed and the last active reference of the requests is completed. Not inactive but free. This is now not possible if we allow the user to not free the persistent requests. To make things worse, if the user now free the buffer that the MPI libraries track as pinned, we are going into potentially unsafe territories.
2. Imagine an MPI implementation that tracks it's objects as integers into a global array of objects, and that allows for reuse of communicator ids. If this implementation remove a communicator that is still references by a persistent requests, and later reuse the same communicator id for a new communicator, any persistent request that was not freed and that uses the old communicator id is now pointing to a valid communicator different from the one that was not the original one associated with the request. 

Yes, we can circumvent these issues in the MPI implementations, at a cost. But why should we when we can address it in the standard in a much friendlier way. And again this has nothing to do with `MPI_Finalize`.

Do you have any evaluation of the implied overhead, or this just common sense ? Because I do have strong indication the required changes are more than just "There may be one additional entry in the internal data structure of the MPI library during the init of a persistent request and some additional specific effort during MPI_COMM_DISCONNECT and MPI_SESSION_FINALIZE.", as it also impacts every persistent communication start.

Last but not least you cannot really claim that your proposal "is backward compatible, because until MPI-4.0 the behavior was undefined.". If the behavior was undefined, anything can be made backward compatible.

**Comment by RolfRabenseifner** [comment]
@bosilca I have many questions to you:
Which MPI library requires that inactive persistent request handles are freed before calling MPI_Finalize?
Which MPI library has significant performance differences between freeing inactive persistent request handles before MPI_Filnalize or not doing this?
Where do you see any sentence in MPI-4.0 that inactive persistent requests must be explicitely freed before
- MPI_Finalize?
- MPI_Session_finalize?
- MPI_Comm_disconnect?
As far as I can see, the only undefined behavior is, whether an explicit MPI_Request_free **after** MPI_Session_finalize or MPI_Comm_disconnect is erroneous or not. Correct?

Your first item clearly says, we should add an advice to users that they should free inactive persistent requests as soon as they do not longer use them.

If you provide such an advice and the forum supports it, then I can include it after the June 7, 2024 virtual meeting.

**Comment by raffenet** [comment]
> it also impacts every persistent communication start.

Can you expand on this?

**Comment by RolfRabenseifner** [comment]
Based on the discussions above and in the [PR823](https://github.com/mpi-forum/mpi-standard/pull/823), 
- I added an **alternative B** to the commented pdf, which does not free inactive persistent requests
  (see pages 1, 100, 475, 525, 1025, 1054, 1057),
- and I added questions about **MPI_Session_finalize** and communicator disconnect (see pages 482 and 524)

for the virtual MPI forum meeting today:
[mpi41-report_Issue710_PR823_update_for 2023-6-7-meeting.pdf](https://github.com/mpi-forum/mpi-standard/files/11676448/mpi41-report_Issue710_PR823_update_for.2023-6-7-meeting.pdf)

**Comment by bkmgit** [comment]
The terms `pending` and `active` mean slightly different things. They may not be exchangeable. Typically understand `pending` as request has been issued and is queued to start (ie. communication may not be in progress, similar to a ringing telephone or signal to initiate telephone ringing has been sent) and `active` is ongoing communication.

**Comment by Wee-Free-Scot** [comment]
@bkmgit The term active is defined in section 2.4.1 of MPI-4.0 -- it means that the starting stage has been done but the completion stage has not (yet). See Figures 2.2 and 2.3 -- state transition into the active state is done by starting the operation, transition from the active state is done by completion of the operation.

Pending is not defined by the MPI Standard, but is used in the MPI Standard as though it is defined.

**Comment by RolfRabenseifner** [comment]
The changes by [PR 823](https://github.com/mpi-forum/mpi-standard/pull/823) are marked with PR823 and also in yellow color on pages 100, 477, 483-484, 527-528, 1027, 1056, 1059, 1091, 1093-1094, 1097 in
[mpi41-report_Issue710_PR823_update5_for-2023-6-14-meeting.pdf](https://github.com/mpi-forum/mpi-standard/files/11739843/mpi41-report_Issue710_PR823_update5_for-2023-6-14-meeting.pdf)

Within this pdf, there is an additional alternative proposal option C, see pages 100 and 483-484.

**Comment by RolfRabenseifner** [comment]
Based on the discussion above, I added several options to the commented pdf of the current Latex version of the [PR 823](https://github.com/mpi-forum/mpi-standard/pull/823) of this issue #710.

The changes by  [PR 823](https://github.com/mpi-forum/mpi-standard/pull/823) are marked with PR823 and also in yellow color on pages 100, 148, 477, 483-484, 527-528, 1027, 1056, 1059, 1091, 1093-1094, 1097
with additional
- option C (green color) on pages 100 and 483-484:
   when a session is finalized, disallowing associated inactive persistent handles;
- option D (cyan color) on page 484:
   AtoT about the optimization of datatype handles;
- options E and F (purple color) on page 148 and 1027:
   datatype handles can be used and freed within the World Model (option E)
   or a session (option F) only if it was created and commited within after
   the World Model or the session was initialized and not yet finalized.

PDF file: [mpi41-report_Issue710_PR823_update6_for-2023-6-19-meeting.pdf](https://github.com/mpi-forum/mpi-standard/files/11787848/mpi41-report_Issue710_PR823_update6_for-2023-6-19-meeting.pdf)

**Comment by RolfRabenseifner** [comment]
@hppritcha and the whole session WG:

The result from the sessions WG meeting today is now added to this [PR 823](https://github.com/mpi-forum/mpi-standard/pull/823). 

The changes by [PR 823](https://github.com/mpi-forum/mpi-standard/pull/823) are marked with PR823 and also in yellow color on pages 100, 148, 473, 475, 477, 479, 483-485, 528, 1027, 1056, 1059, 1091, 1094-1095, 1097-1099
with additional
- option C on pages 100 and 484 (green color)
   (when a session is finalized, disallowing associated inactive persistent handles)

PDF file: [mpi41-report_Issue710_PR823_update9_for-2023-6-21-meeting.pdf](https://github.com/mpi-forum/mpi-standard/files/11793445/mpi41-report_Issue710_PR823_update9_for-2023-6-21-meeting.pdf)

It will be the input for the reading at the virtual MPI forum meeting on June 21, 2023.

**Comment by RolfRabenseifner** [comment]
Here the protocol from the Virtual MPI forum meeting, June 21, 2023:
[mpi41-report_Issue710_PR823_update9_for-2023-6-21-meeting-protocol.pdf](https://github.com/mpi-forum/mpi-standard/files/11834209/mpi41-report_Issue710_PR823_update9_for-2023-6-21-meeting-protocol.pdf)

**Comment by RolfRabenseifner** [comment]
The result from the sessions WG meeting on June 19, 2023 and the virtual MPI forum meeting on June 21, 2023 is now added to this [PR 823](https://github.com/mpi-forum/mpi-standard/pull/823).

Commented PDF file: [mpi41-report_Issue710_PR823_update10.pdf](https://github.com/mpi-forum/mpi-standard/files/11835937/mpi41-report_Issue710_PR823_update10.pdf)

The changes by this [PR 823](https://github.com/mpi-forum/mpi-standard/pull/823) are marked with PR823 and also in yellow color on pages 100, 148, 477, 483-484, 527-528, 1027, 1056, 1059, 1091, 1094-1095, 1097-1099 with additional
- option C on pages 100 and 484 (light yellow / green color)
  - without option C: same behavior for inactive persistent request for
                                  MPI_COMM_DISCONNECT, MPI_FINALIZE and MPI_SESSION_FINALIZE.
  - with option C: when a session is finalized, disallowing associated inactive persistent handles
- options A or B on page 148: solving the problem about datatype handles, see page 484
  - with option A: Advice to implementors: relation beteen MPI_TYPE_COMMIT nd session initialization and finalization.
  - with option B: Same as advice to users: hint that MPI libraries may have problems with the optimization of datatypes in case of session initialization and finalization.

It will be the input for the errata reading and voting at the MPI forum meeting on July 10-13, 2023.
Before voting, two decisions are needed (e.g., by straw votes):
- between **option A** and **option B**,
- between **without option C** and **with option C**.

**Comment by schulzm** [comment]
As for Option C: while I understand the motivation to make a change here to fix something that we don't like in the world model, I would strongly argue against this: this would create a difference between the default "world" session and all other sessions. This will not only cause issues for tools, as discussed in the last call, it will also cause problems for any library that has that behavior - those could not be used safely in a session model, but this is hard to determine for the end user (and the library itself would not be aware of the fact that it is used from sessions). Also, the property of not removing persistent requests at the end would not be an unexpected behavior for a library that reuses persistent requests across library calls.

**Comment by RolfRabenseifner** [comment]
The result from the sessions WG meeting on June 19, 2023 and the virtual MPI forum meeting on June 21, 2023 is now added to this [PR 823](https://github.com/mpi-forum/mpi-standard/pull/823), plus a few additional English corrections  (June 22).

Commented PDF file: [mpi41-report_Issue710_PR823_update11.pdf](https://github.com/mpi-forum/mpi-standard/files/11866140/mpi41-report_Issue710_PR823_update11.pdf)

The changes by this [PR 823](https://github.com/mpi-forum/mpi-standard/pull/823) are marked with PR823 and also in yellow color on pages 100, 148, 477, 483-484, 527-528, 1027, 1056, 1059, 1091, 1094-1095, 1097-1099 with additional
- option C on pages 100 and 484 (light yellow / green color)
  - without option C: same behavior for inactive persistent request for
                                  MPI_COMM_DISCONNECT, MPI_FINALIZE and MPI_SESSION_FINALIZE.
  - with option C: when a session is finalized, disallowing associated inactive persistent handles
- options A or B on page 148: solving the problem about datatype handles, see page 484
  - with option A: Advice to implementors: relation beteen MPI_TYPE_COMMIT nd session initialization and finalization.
  - with option B: Same as advice to users: hint that MPI libraries may have problems with the optimization of datatypes in case of session initialization and finalization.

It will be the input for the errata reading and voting at the MPI forum meeting on July 10-13, 2023.
Before voting, two decisions are needed (e.g., by straw votes):
- between **option A** and **option B**,
- between **without option C** and **with option C**.

**Comment by Wee-Free-Scot** [comment]
@schulzm makes three points against option C; with extensive editorializing, these are:
1) it creates a difference between the "world session" and all other sessions -- this relies on the existence of lazy users who do not (wish to) clean up handles from a library
2) the difference between world and other sessions causes problems for tools -- this is predicated on the existence of tools that intercept MPI_INIT & MPI_INIT_THREAD & MPI_FINALIZE and choose to implement them using PMPI_SESSION_INIT & PMPI_SESSION_FINALIZE, rather than the more traditional route of using PMPI_INIT & PMPI_INIT_THREAD & PMPI_FINALIZE 
3) the difference between world and other sessions causes problems for libraries -- this hinges on the existence of libraries that do not initialize either the world model or a session themselves, but use MPI objects passed to them from outside the library without being told whether those objects are derived from the world model or the sessions model

Rebuttals for these three objections might use the following arguments:
1) this is the intent of the option and will be seen by those in favor of option C as A Good Thing (tm)
2) this is a strange implementation choice for the tool; there is no need to incur this implementation difficultly unless the tool writer wishes to incur it
3) this violates the intent of the sessions model, which invites each library to initialize its own session and not to share MPI objects externally (neither "in" nor "out"); also handles that are created outwith the library are not the responsibility of the library (in this case, the library is a distraction) and libraries that create handles the tie-up resources but fail to release them are not well written and are less usable than ones that clean up after themselves

How far should we go out of our way to support bad user code?

**Comment by schulzm** [comment]
To 2) - this is not bad code or a bad design decision - tools do subset COMM_WORLD and mask that upwards to the application and one of the use cases for sessions was to support tools with extra processes, e.g., in overlay networks. Hence, this will not be an usual case for tools - or rather we are preventing legitimate tool use cases

To 3) - even a session-ized library will have to (for quite some time) rely on existing other libraries and hence falls into this category, hence, again not bad code: we would basically be saying, if you want to switch to sessions you cannot rely on any existing MPI library/code anymore, you have to write everything from scratch, unless you know exactly what is in this library. This will not help us to promote the use of sessions - nobody will want to give up using the existing MPI ecosystem.

**Comment by RolfRabenseifner** [comment]
Protocol from the formal reading, July 10, 2023:
[mpi41-report_Issue710_PR823_update12-from-reading.pdf](https://github.com/mpi-forum/mpi-standard/files/12008719/mpi41-report_Issue710_PR823_update12-from-reading.pdf)

**Comment by RolfRabenseifner** [comment]
Final version for no-no-vote: [mpi41-report_Issue710_PR823_update14-for-voting.pdf](https://github.com/mpi-forum/mpi-standard/files/12015270/mpi41-report_Issue710_PR823_update14-for-voting.pdf)

**Comment by wesbland** [comment]
This passed a [no-no vote](https://mpi-forum.org/meetings/2023/07/votes).
<table>
<tr>
    <td> Yes
    <td> No
    <td> Abstain
<tr>
    <td> 30
    <td> 0
    <td> 4
</table>

**Comment by wesbland** [comment]
This passed a [errata vote](https://mpi-forum.org/meetings/2023/07/votes).
<table>
<tr>
    <td> Yes
    <td> No
    <td> Abstain
<tr>
    <td> 27
    <td> 2
    <td> 5
</table>

**Comment by hzhou** [comment]
I am very surprised that one of the conclusions of this issue is -- users do not need free a persistent request that they explicitly created. This is just odd with no precedence. I am a bit upset that this gets sneaked in as errata. It is one thing that the previous text **neglected** to mention the requirement, but it is another thing to issue an errata to clarify that the neglect was **intentional**.

I guess I am ahead of myself -- Is my interpretation of the errata correct? And what is the logic that allows explicit MPI object creation without corresponding explicit free/destroy?

@RolfRabenseifner 

PS: sorry to comment on an already closed issue.

**Comment by Wee-Free-Scot** [comment]
I created a new issue to capture the ongoing discussion of this: https://github.com/mpi-forum/mpi-issues/issues/732

**Comment by hzhou** [comment]
Thanks @Wee-Free-Scot for creating new tracking issue. 

Separately, @raffenet and I think it is questionable for the current errata procedures. Errata by common sense should be a correction of *non-controversial* mistakes. The errata in question has significant negative votes -- 2 No and 5 Abstain. Thus, in my opinion, should not qualify as an errata. We think passing errata should have a much higher bar than other votes. For example, I think it is reasonable to expect "No-No" for errata votes.

**Comment by hzhou** [comment]
@wgropp What do you think https://github.com/mpi-forum/mpi-issues/issues/710#issuecomment-1636147221 ?
